### ISSUE TITLE
Possible false positive thread safety violation

### ISSUE THREAD

**Issue by whymarrh** [facebook/infer#785]
The following Java class is flagged as having a thread safety issue. I think this may be a false positive generated by the recently introduced [RacerD](http://fbinfer.com/docs/racerd.html). Infer suggests that `time()` reads `time` without synchronization but the read is guarded by `lock`.

```java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.LongFunction;

public final class Clock {
    private final Lock lock = new ReentrantLock();

    private long time = 0L;

    public long time() {
        lock.lock();
        try {
            return time;
        } finally {
            lock.unlock();
        }
    }

    public <T> T tick(final LongFunction<T> ticker) {
        lock.lock();
        try {
            time = time + 1;
            return ticker.apply(time);
        } finally {
            lock.unlock();
        }
    }
}
```

The issue that Infer points out:

```
$ infer -- javac Clock.java 
Capturing in javac mode...
Found 1 source file to analyze in /Projects/rxb/infer-out
Starting analysis...

legend:
  "F" analyzing a file
  "." analyzing a procedure

F...

Found 1 issue

Clock.java:13: error: THREAD_SAFETY_VIOLATION
  Read/Write race. Non-private method `Clock.time` reads without synchronization from `&this.Clock.time`. Potentially races with writes in method `Object Clock.tick(LongFunction)`.
 Reporting because this access may occur on a background thread.
  11.           lock.lock();
  12.           try {
  13. >             return time;
  14.           } finally {
  15.               lock.unlock();


Summary of the reports

  THREAD_SAFETY_VIOLATION: 1
```

The issue can be worked around by simply assigning a temporary variable and returning that:

```java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.LongFunction;

public final class Clock {
    private final Lock lock = new ReentrantLock();

    private long time = 0L;

    public final long time() {
        lock.lock();
        try {
            final long x = time;
            return x;
        } finally {
            lock.unlock();
        }
    }

    public final <T> T tick(final LongFunction<T> ticker) {
        lock.lock();
        try {
            time = time + 1;
            return ticker.apply(time);
        } finally {
            lock.unlock();
        }
    }
}
```

Oracle Java version 1.8:

```
java version "1.8.0_152"
Java(TM) SE Runtime Environment (build 1.8.0_152-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.152-b16, mixed mode)
```

Infer version 0.13.0:

```
Infer version v0.13.0-ddda04c9
Copyright 2009 - present Facebook. All Rights Reserved.
Usage: infer command [options]
See `infer --help` for more information.
```

Infer was built locally on macOS Sierra 10.12.6 using the following steps:

```sh
brew pull 19646
brew cask install caskroom/versions/java8
brew install infer # to get deps
git clone git@github.com:facebook/infer
cd infer
opam init --no-setup
./build-infer.sh --yes java
```

**Comment by sblackshear** [comment]
Thanks for the detailed report! I see the problem and am working on a fix.

**Comment by sblackshear** [comment]
Should be fixed by 4ee4ebb42aa3dced9e191cdf05742ec774ffce18.

**Comment by whymarrh** [comment]
Thanks!

**Comment by whymarrh** [comment]
@sblackshear I'm still seeing this with 0.13.1, should that be the case?

**Comment by jvillard** [comment]
@whymarrh the fix appeared after 0.13.0, and 0.13.1 is just a re-release of 0.13.1 to fix a build issue.

I'll try to make a new release soonish.

**Comment by whymarrh** [comment]
Ah, yes. Thanks for the clarification!
